### PR TITLE
Full support of sparse AD for NLS models

### PR DESCRIPTION
**Pull request by amontoison**
@tmigot Can you help me with it?

### CODE REVIEW

**Comment by tmigot** [comment]
@amontoison I keep this in my list. Could you rebase the branch?

**Comment by amontoison** [comment]
> @amontoison I keep this in my list. Could you rebase the branch?

@tmigot Done

**Comment by tmigot** [comment]
@amontoison I took care of the ADNLPModel-stuff, just rest to fix `hess_structure_residual!` and `hess_coord_residual!` for sparse backend.

File: src/sparse_hessian.jl [hunk 1]
```diff
@@ -294,3 +294,44 @@ function NLPModels.hess_coord!(
...
+  b::Union{SparseADHessian, SparseReverseADHessian},
+  nls::AbstractADNLSModel,
+  rows::AbstractVector{<:Integer},
+  cols::AbstractVector{<:Integer},
+)
+  function objective(x)
+    F = get_F(nls, b)
+    Fx = F(x)
+    return dot(Fx, Fx) / 2
+  end
```
Comment (tmigot) [inline src/sparse_hessian.jl:308]: This is not the function we want to differentiate, but `dot(Fx, v)/2` for a given `v`.
I guess it should look a lot like the Hessian of the Lagrangian for obj_weight=0.

File: src/sparse_hessian.jl [hunk 2]
```diff
@@ -294,3 +294,44 @@ function NLPModels.hess_coord!(
...
+  nls::AbstractADNLSModel,
+  x::AbstractVector,
+  v::AbstractVector,
+  vals::AbstractVector,
+)
+  function objective(x)
+    F = get_F(nls, b)
+    Fx = F(x)
+    return dot(Fx, Fx) / 2
+  end
```
Comment (tmigot) [inline src/sparse_hessian.jl:334]: same as above

File: test/sparse_hessian_nls.jl [hunk 3]
```diff
@@ -0,0 +1,44 @@
+list_sparse_hess_backend = (
+  (ADNLPModels.SparseADHessian, Dict()),
```
Comment (tmigot) [inline test/sparse_hessian_nls.jl:2]: According to the implementation above `SparseReverseADHessian` should work too.

Comment (amontoison) [reply test/sparse_hessian_nls.jl:2]: We also forgot it in `test/sparse_hessian.jl`. I will update both files.

File: src/sparse_hessian.jl (hunk 1 above)
Comment (amontoison) [reply src/sparse_hessian.jl:308]: The objective of an NLS problem is not `0.5 * \|F(x)\|^2` ???

Comment (tmigot) [reply src/sparse_hessian.jl:310]: `hess_residual` is the "Hessian" of the residual map not the objective function

**Comment by amontoison** [comment]
> @amontoison I took care of the ADNLPModel-stuff, just rest to fix `hess_structure_residual!` and `hess_coord_residual!` for sparse backend.

@tmigot I think we need to create the backend for `residual_hessian` with different arguments (`f(x) = 0` and `c(x) = F(x)`).
You probably know how to fix that.

**Comment by amontoison** [comment]
@tmigot Tu sais à quoi ça sert ces lignes?
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/commit/d37a9a1c27069eb2741bf73686d08ca5d27501d0#diff-a1616586d29af268a473ea5c9c5ea6d09aa5f0de50e914a9d0bedc8d4f7bb59eR92-R99

**Comment by tmigot** [comment]
> @tmigot Tu sais à quoi ça sert ces lignes? [d37a9a1#diff-a1616586d29af268a473ea5c9c5ea6d09aa5f0de50e914a9d0bedc8d4f7bb59eR92-R99](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/commit/d37a9a1c27069eb2741bf73686d08ca5d27501d0#diff-a1616586d29af268a473ea5c9c5ea6d09aa5f0de50e914a9d0bedc8d4f7bb59eR92-R99)

C'est pour couvrir le cas où le "backend" fourni par l'utilisateur est un NLPModel. Mais définir cette fonction
```
function get_residual_nnzj(nlp::AbstractNLPModel, nvar, nequ)
  nlp.nls_meta.nnzj
end
```
devrait être suffisant je pense.

**Comment by amontoison** [comment]
@tmigot All tests passed for Hessian but we probably need to update the API for `jth_hess_residual`.

**Comment by tmigot** [comment]
Thanks for fixing it. I think the implementation in AD&NLPModels of jth_residual_hess is not clear so I removed the additional implementation (https://github.com/JuliaSmoothOptimizers/NLPModels.jl/issues/466). It should work now

**Review by tmigot — APPROVED** [review]
Thanks @amontoison ! LGTM

**Pull request merged**